precommit.py

```diff
@@ -0,0 +1,142 @@
+        "If not set, an exception is raised "
+        "if any of the files do not conform to the style guide.",
+        action="store_true",
+    )
+    parser.add_argument(
+        "--select",
+        help="If set, only the selected steps are executed. "
+        "This is practical if some of the steps failed and you want to fix them in isolation.",
+        nargs="+",
+        choices=[value.value for value in Step],
+    )
+    parser.add_argument(
+        "--skip",
+        help="If set, skips the specified steps. "
+        "This is practical if some of the steps passed and "
+        "you want to fix the remainder in isolation.",
+        nargs="+",
+        choices=[value.value for value in Step],
+    )
+
+    args = parser.parse_args()
+
+    overwrite = bool(args.overwrite)
+
+    selects = (
+        [Step(value) for value in args.select]
+        if args.select is not None
+        else [value for value in Step]
+    )
+    skips = [Step(value) for value in args.skip] if args.skip is not None else []
+
+    repo_root = pathlib.Path(__file__).parent
+
+    if Step.BLACK in selects and Step.BLACK not in skips:
+        print("Black'ing...")
+        # fmt: off
+        black_targets = [
+            "crosshair",
+            "precommit.py",
+            "setup.py"
+        ]
+        # fmt: on
+
+        if overwrite:
+            subprocess.check_call(["black"] + black_targets, cwd=str(repo_root))
+        else:
+            subprocess.check_call(
+                ["black", "--check"] + black_targets, cwd=str(repo_root)
+            )
+    else:
+        print("Skipped black'ing.")
+
+    if Step.FLAKE8 in selects and Step.FLAKE8 not in skips:
+        print("Flake8'ing...")
+        # fmt: off
+        subprocess.check_call(
+            [
+                "flake8", "crosshair", "--count", "--select=E9,F63,F7,F82", "--show-source",
+                "--statistics"
+            ],
+            cwd=str(repo_root)
+        )
+        # fmt: on
+    else:
+        print("Skipped flake8'ing.")
+
+    if Step.PYDOCSTYLE in selects and Step.PYDOCSTYLE not in skips:
+        print("Pydocstyle'ing...")
+        subprocess.check_call(
+            [
+                "pydocstyle",
+                "--ignore=D1,D203,D212",
+                r"--match=.*(?<!_test).py$",
+                r"--match-dir=(?!/examples/)",
+                "crosshair",
+            ],
+            cwd=str(repo_root),
+        )
+    else:
+        print("Skipped pydocstyle'ing.")
+
+    if Step.TEST in selects and Step.TEST not in skips:
+        print("Testing...")
+        env = os.environ.copy()
+        env["ICONTRACT_SLOW"] = "true"
+
+        # fmt: off
+        subprocess.check_call(
+            [
+                "coverage", "run",
+                "--source", "crosshair",
+                "--omit=__init__.py"
+                "--omit=*_test.py",
+                "-m", "pytest",
+                "--doctest-modules"
```

Comment:
I **think** this will run the docstring doctests and potentially we could add  `--doctest-glob="*.md"` to cover the doctest target below.

I'll merge and you can follow up if you like.

Reply:
Actually, I did the separation on purpose since the tests take quite long to run. I'll add a comment in the script.

Reply:
SGTM. The doctest coverage right now doesn't make for a great "smoke test," but we'll work on it!